plugins/slack_app/app/models/slack_message.rb

```diff
@@ -0,0 +1,116 @@
+class SlackMessage
+  def initialize(deploy)
+    @deploy = deploy
+  end
+
+  def deliver
+    url = DeployResponseUrl.find_by(deploy_id: @deploy.id).try(:response_url)
+    Faraday.new(url).post do |req|
+      req.headers['Content-Type'] = 'application/json'
+      req.body = ActiveSupport::JSON.encode message_body
+    end if url.present?
+  end
+
+  def message_body
+    ret = if @deploy.waiting_for_buddy?
+      waiting_for_buddy_body
+    elsif @deploy.failed? || @deploy.errored?
+      failed_body
+    elsif @deploy.succeeded?
+      succeeded_body
+    else
+      running_body
+    end
+    ret['response_type'] = 'in_channel'
+    ret
+  end
+
+  private
+
+  def waiting_for_buddy_body
+    {
+      text: title_str,
+      attachments: [
+        {
+          text: pr_str,
+          callback_id: @deploy.id,
+          attachment_type: 'default',
+          actions: [
+            button(':+1: Approve', 'yes')
+          ]
+        }
+      ]
+    }
+  end
+
+  def running_body
+    {
+      text: title_str,
+      attachments: [
+        {
+          text: pr_str + "\nDeploying…",
+          attachment_type: 'default',
+          color: 'warning'
+        }
+      ]
+    }
+  end
+
+  def failed_body
+    {
+      text: ':x: ' + title_str('failed to deploy'),
+    }
+  end
+
+  def succeeded_body
+    {
+      text: ':tada: ' + title_str('successfully deployed')
+    }
+  end
+
+  def title_str(ended = nil)
+    if @deploy.buddy
+      owner = SlackIdentifier.find_by_user_id(@deploy.user.id)
```

Comment:
yeah that's ugly :D
... maybe there is something nice around displaying json errors that we can
do there .... good enough for now ...

On Sun, Jul 17, 2016 at 4:02 PM, Ben Straub <notifications@github.com>
wrote:

> In plugins/slack_app/app/models/slack_message.rb
> <https://github.com/zendesk/samson/pull/1130#discussion_r71091625>:
>
> > +
> > +  def failed_body
> > +    {
> > +      text: ':x: ' + title_str('failed to deploy'),
> > +    }
> > +  end
> > +
> > +  def succeeded_body
> > +    {
> > +      text: ':tada: ' + title_str('successfully deployed')
> > +    }
> > +  end
> > +
> > +  def title_str(ended = nil)
> > +    if @deploy.buddy
> > +      owner = SlackIdentifier.find_by_user_id(@deploy.user.id)
>
> Hmm. Here's what it looks like in Slack if one of the command or interact
> endpoints throws:
>
> [image: image]
> <https://cloud.githubusercontent.com/assets/39902/16903600/ca6c4946-4c37-11e6-869c-f705aeccace4.png>
>
> I don't really like that experience. And I'm not sure how to replace it,
> so I'm doing a lot of error checking.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zendesk/samson/pull/1130/files/d008e855296f051e83d429a560e4e921b94fd917#r71091625>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAsZ4w6HhilurY9u-3fmJAqRhbK4saCks5qWrSXgaJpZM4JIRHk>
> .
>